java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -563,15 +567,18 @@ public void visitNewClass(NewClassTree tree) {
       scan(classBody);
       registerType(tree, anonymousClassType);
     } else {
-      registerType(tree, getType(tree.identifier()));
+      JavaType resolutionType = resolution.type();
+      JavaType type = resolutionType != null && resolutionType != symbols.voidType ? resolutionType : getType(tree.identifier());
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)